packages/dds/task-manager/src/taskManager.ts

```diff
@@ -193,6 +203,26 @@ export class TaskManager extends SharedObject<ITaskManagerEvents> implements ITa
             this.removeClientFromQueue(taskId, clientId);
         });
 
+        this.opWatcher.on("complete", (taskId: string, clientId: string, local: boolean, messageId: number) => {
+            if (runtime.connected && local) {
+                const pendingOp = this.latestPendingOps.get(taskId);
+                assert(pendingOp !== undefined, "Unexpected op");
+                // TODO: check below comment and stuff, see if applicable
+                // Need to check the id, since it's possible to complete multiple times before the acks
+                if (messageId === pendingOp.messageId) {
+                    assert(pendingOp.type === "complete", "Unexpected op type");
+                    // Delete the pending, because we no longer have an outstanding op
+                    this.latestPendingOps.delete(taskId);
+                }
+            }
+
+            if (this.taskQueues.has(taskId)) {
```

Comment:
Thinking about this more, there's probably more we need to do to track the pending "complete".  Very similar to how we have to track "clear" specially for map/directory, I think we maybe need to ignore/drop incoming volunteer ops if we know a "complete" ack is coming our way (otherwise e.g. the local client will get a second "completed" event here for their ack, if some other client volunteers between the sending and the ack):

* ClientA volunteer()
* (ack ClientA volunteer)
* ClientA sees "assigned" event
* ClientB volunteer()
* ClientA complete()
* ClientA sees a "completed" event
* (ack ClientB volunteer) -- ClientA creates a new taskQueue with ClientB in it
* (ack ClientA complete)
* Both ClientA and ClientB see a "completed" event

Reply:
Got it. I reworked the logic and added a test for this scenario in the latest commit (the test failed on the previous commit)